test/php/gutenberg-coding-standards/Gutenberg/Sniffs/Commenting/FunctionCommentSinceTagSniff.php

```diff
@@ -48,90 +91,444 @@ public function process( File $phpcsFile, $stackPtr ) {
+				$phpcs_file->addError( $missing_since_tag_error_message, $since_tag_token, $violation_code );
+				continue;
+			}
+
+			$version_value = $tokens[ $version_value_token ]['content'];
+
+			if ( static::validate_version( $version_value ) ) {
+				continue;
+			}
+
+			$phpcs_file->addError(
+				'Invalid @since version value for the "%s()" hook function: "%s". Version value must be greater than or equal to 0.0.1.',
+				$version_value_token,
+				'InvalidHookSinceTagVersionValue',
+				array(
+					$hook_function,
+					$version_value,
+				)
+			);
 		}
+	}
 
-		$missing_since_tag_error_message = sprintf( '@since tag is missing for the "%s()" function.', $function_name );
+	/**
+	 * Processes a token representing an object-oriented programming structure
+	 * like a class, interface, trait, or enum to check for a missing `@since` tag in its docblock.
+	 *
+	 * @param File $phpcs_file    The file being scanned.
+	 * @param int  $stack_pointer The position of the OO token in the stack.
+	 */
+	protected function process_oo_token( File $phpcs_file, $stack_pointer ) {
+		$tokens                 = $phpcs_file->getTokens();
+		$token_type             = static::$oo_tokens[ $tokens[ $stack_pointer ]['code'] ]['name'];
+		$capitalized_token_type = ucfirst( $token_type );
 
-		// All these tokens could be present before the docblock.
-		$tokens_before_the_docblock = array(
-			T_PUBLIC,
-			T_PROTECTED,
-			T_PRIVATE,
-			T_STATIC,
-			T_FINAL,
-			T_ABSTRACT,
-			T_WHITESPACE,
+		$token_name                      = ObjectDeclarations::getName( $phpcs_file, $stack_pointer );
+		$missing_since_tag_error_message = sprintf(
+			'@since tag is missing for the "%s" %s.',
+			$token_name,
+			$token_type
 		);
 
-		$doc_block_end_token = $phpcsFile->findPrevious( $tokens_before_the_docblock, ( $stackPtr - 1 ), null, true, null, true );
-		if ( ( false === $doc_block_end_token ) || ( T_DOC_COMMENT_CLOSE_TAG !== $tokens[ $doc_block_end_token ]['code'] ) ) {
-			$phpcsFile->addError( $missing_since_tag_error_message, $stackPtr, 'MissingSinceTag' );
+		$violation_code = 'Missing' . $capitalized_token_type . 'SinceTag';
+
+		$docblock = static::find_docblock( $phpcs_file, $stack_pointer );
+		if ( false === $docblock ) {
+			$phpcs_file->addError( $missing_since_tag_error_message, $stack_pointer, $violation_code );
 			return;
 		}
 
-		// The sniff intentionally doesn't check if the docblock has a valid open tag.
-		// Its only job is to make sure that the @since tag is present and has a valid version value.
-		$doc_block_start_token = $phpcsFile->findPrevious( Tokens::$commentTokens, ( $doc_block_end_token - 1 ), null, true, null, true );
-		if ( false === $doc_block_start_token ) {
-			$phpcsFile->addError( $missing_since_tag_error_message, $stackPtr, 'MissingSinceTag' );
+		list( $doc_block_start_token, $doc_block_end_token ) = $docblock;
+
+		$version_tags = static::parse_since_tags( $phpcs_file, $doc_block_start_token, $doc_block_end_token );
```

Comment:
This logic to get the `@since` tags and then check the version values seems to be repeated a few times across this class. Maybe it can be consolidated in a single method? Maybe even the logic above to find the docblock and return a error if one is not found can also be consolidated?

Reply:
I agree. I've refactored the code to avoid duplicates.